plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php

```diff
@@ -0,0 +1,210 @@
+		$defaults = array(
+			'isDescendentOfSingleProductBlock' => false,
+		);
+
+		return wp_parse_args( $attributes, $defaults );
+	}
+
+	/**
+	 * Extra data passed through from server to client for block.
+	 *
+	 * @param array $attributes  Any attributes that currently are available from the block.
+	 *                           Note, this will be empty in the editor context when the block is
+	 *                           not in the post content on editor load.
+	 */
+	protected function enqueue_data( array $attributes = [] ) {
+		parent::enqueue_data( $attributes );
+		$this->asset_data_registry->add( 'isBlockifiedAddToCart', Features::is_enabled( 'blockified-add-to-cart' ) );
+	}
+
+	/**
+	 * Render the block.
+	 *
+	 * @param array    $attributes Block attributes.
+	 * @param string   $content Block content.
+	 * @param WP_Block $block Block instance.
+	 *
+	 * @return string | void Rendered block output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		global $product;
+
+		$post_id = $block->context['postId'];
+
+		if ( ! isset( $post_id ) ) {
+			return '';
+		}
+
+		$previous_product = $product;
+		$product          = wc_get_product( $post_id );
+		if ( ! $product instanceof \WC_Product ) {
+			$product = $previous_product;
+
+			return '';
+		}
+
+		$is_external_product_with_url = $product instanceof \WC_Product_External && $product->get_product_url();
+		ob_start();
+
+		/**
+		 * Trigger the single product add to cart action for each product type.
+		 *
+		 * @since 9.7.0
+		 */
+		do_action( 'woocommerce_' . $product->get_type() . '_add_to_cart' );
+
+		$product_html = ob_get_clean();
+
+		if ( ! $product_html ) {
+			$product = $previous_product;
+
+			return '';
+		}
+		$product_name = $product->get_name();
+
+		$parsed_attributes                     = $this->parse_attributes( $attributes );
+		$is_descendent_of_single_product_block = $parsed_attributes['isDescendentOfSingleProductBlock'];
+
+		if ( ! $is_external_product_with_url ) {
+			$product_html = $this->add_is_descendent_of_single_product_block_hidden_input_to_product_form( $product_html, $is_descendent_of_single_product_block );
+		}
+
+		$classes_and_styles = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes, array(), array( 'extra_classes' ) );
+
+		$product_classname = $is_descendent_of_single_product_block ? 'product' : '';
+
+		$classes = implode(
+			' ',
+			array_filter(
+				array(
+					'wp-block-add-to-cart-with-options wc-block-add-to-cart-with-options',
+					esc_attr( $classes_and_styles['classes'] ),
+					esc_attr( $product_classname ),
+					'wc-block-add-to-cart-with-options--input',
+				)
+			)
+		);
+
+		$wrapper_attributes = get_block_wrapper_attributes(
+			array(
+				'class' => $classes,
+				'style' => esc_attr( $classes_and_styles['styles'] ),
+			)
+		);
+
+		$form = sprintf(
+			'<div %1$s %2$s>%3$s</div>',
+			$wrapper_attributes,
+			'',
+			$product_html
+		);
```

Comment:
We should try to avoid markup manipulation as much as possible. Let's use the HTML API.
Not a blocker, but something to change in the short time, imo

Reply:
For reference, issue created here: https://github.com/woocommerce/woocommerce/issues/53282.